File: app/src/main/java/com/zacharee1/sliderpreferenceembedded/SliderPreferenceEmbedded.java
package com.zacharee1.sliderpreferenceembedded;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.preference.Preference;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

import org.adw.library.widgets.discreteseekbar.DiscreteSeekBar;

public class SliderPreferenceEmbedded extends Preference {
    private View mView;
    private OnPreferenceChangeListener mListener;
    private OnViewCreatedListener mViewListener;

    private int mProgress = -1;
    private int mMaxProgress = -1;
    private int mMinProgress = -1;
    private int mXmlProgress = 0;

    private float mScale = 1.0F; //doesn't work for the popup view
    private int mPrecision = 2;

    private boolean popupEnabled = true;
    private boolean textEnabled = false;

    private String mFormat = null;

    public DiscreteSeekBarText seekBar;

    @TargetApi(21)
    public SliderPreferenceEmbedded(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public SliderPreferenceEmbedded(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public SliderPreferenceEmbedded(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.SliderPreferenceEmbedded,
                0, 0);

        try {
            mMaxProgress = a.getInteger(R.styleable.SliderPreferenceEmbedded_seek_max, -1);
            mMinProgress = a.getInteger(R.styleable.SliderPreferenceEmbedded_seek_min, -1);
            mXmlProgress = a.getInteger(R.styleable.SliderPreferenceEmbedded_default_progress, -1);
            mFormat = a.getString(R.styleable.SliderPreferenceEmbedded_format);
            popupEnabled = a.getBoolean(R.styleable.SliderPreferenceEmbedded_show_popup, popupEnabled);
            textEnabled = a.getBoolean(R.styleable.SliderPreferenceEmbedded_show_text, textEnabled);
            mScale = a.getFloat(R.styleable.SliderPreferenceEmbedded_scale, mScale);
            mPrecision = a.getInteger(R.styleable.SliderPreferenceEmbedded_precision, mPrecision);
        } finally {
            a.recycle();
        }
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        setLayoutResource(R.layout.pref_view_embedded);
        setWidgetLayoutResource(R.layout.slider_pref_view);

        mView = super.onCreateView(parent);

        mProgress = (mProgress == -1 ? getSavedProgress() : mProgress);
        mMaxProgress = (mMaxProgress == -1 ? 100 : mMaxProgress);
        mMinProgress = (mMinProgress == -1 ? 0 : mMinProgress);

        seekBar = mView.findViewById(R.id.seekbar_view);
        seekBar.setScale(mScale);
        seekBar.setPrecision(mPrecision);
        seekBar.setMin(mMinProgress);
        seekBar.setMax(mMaxProgress);
        seekBar.setIndicatorFormatter(mFormat);
        seekBar.setProgress(mProgress);
        seekBar.setTextIndicatorEnabled(textEnabled);
        seekBar.setPopupIndicatorEnabled(popupEnabled);

        seekBar.setOnProgressChangeListener(new DiscreteSeekBarText.OnProgressChangeListener() {
            @Override
            public void onProgressChanged(DiscreteSeekBar seekBar, int value, boolean fromUser) {
                setProgressWithoutBar(value);
                if (mListener != null) mListener.onPreferenceChange(SliderPreferenceEmbedded.this, value);
            }

            @Override
            public void onStartTrackingTouch(DiscreteSeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(DiscreteSeekBar seekBar) {
                setProgressWithoutBar(seekBar.getProgress());
                if (mListener != null) mListener.onPreferenceChange(SliderPreferenceEmbedded.this, seekBar.getProgress());
            }
        });

        if (mViewListener != null) mViewListener.viewCreated(this);
        return mView;
    }

    @Override
    protected void onBindView(View view) {
        super.onBindView(view);

        if (mViewListener != null) mViewListener.viewBound(this);
    }

    @Override
    public void setOnPreferenceChangeListener(OnPreferenceChangeListener onPreferenceChangeListener) {
        mListener = onPreferenceChangeListener;
    }

    public void setOnViewCreatedListener(OnViewCreatedListener listener) {
        mViewListener = listener;
    }

    public void setProgress(int progress) {
        if (seekBar != null) seekBar.setProgress(progress);

        setProgressWithoutBar(progress);
    }

    public void setMax(int max) {
        mMaxProgress = max;
        if (seekBar != null) seekBar.setMax(max);
    }

    public void setMin(int min) {
        mMinProgress = min;
        if (seekBar != null) seekBar.setMin(min);
    }

    public void setPopupEnabled(boolean enabled) {
        popupEnabled = enabled;
        seekBar.setPopupIndicatorEnabled(enabled);
    }

    public void setTextEnabled(boolean enabled) {
        textEnabled = enabled;
        seekBar.setTextIndicatorEnabled(enabled);
    }

    public void setScale(float scale) {
        mScale = scale;
        seekBar.setScale(scale);
    }

    public int getMax() {
        return mMaxProgress;
    }

    public int getMin() {
        return mMinProgress;
    }

    public int getProgress() {
        return mProgress;
    }

    public boolean getPopupEnabled() {
        return seekBar.getPopupIndicatorEnabled();
    }

    public boolean getTextEnabled() {
        return seekBar.getTextIndicatorEnabled();
    }

    private void setProgressWithoutBar(int progress) {
        setProgressState(progress);
        saveProgress(progress);
    }

    private void saveProgress(int progress) {
        getSharedPreferences().edit().putInt(getKey(), progress).apply();
    }

    private void setProgressState(int progress) {
        mProgress = progress;
    }

    private int getSavedProgress() {
        return getSharedPreferences().getInt(getKey(), mXmlProgress);
    }

    public interface OnViewCreatedListener {
        void viewCreated(SliderPreferenceEmbedded preferenceEmbeddedNew);
        void viewBound(SliderPreferenceEmbedded preferenceEmbeddedNew);
    }
}

File: app/src/main/java/com/zacharee1/sliderpreferenceembedded/SliderPreferenceEmbedded.kt
package com.zacharee1.sliderpreferenceembedded

import android.content.Context
import android.content.res.TypedArray
import android.preference.Preference
import android.preference.PreferenceManager
import android.support.v4.graphics.ColorUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.rey.material.widget.Slider
import java.util.*

class SliderPreferenceEmbedded(context: Context, attrs: AttributeSet) : DefaultValuePreference(context, attrs) {
    var viewListener: OnViewCreatedListener? = null

    lateinit var view: View

    var seekBar: DiscreteSeekBarText

    private var listener: Preference.OnPreferenceChangeListener? = null

    init {
        seekBar = DiscreteSeekBarText(context)

        layoutResource = R.layout.pref_view_embedded
        widgetLayoutResource = R.layout.slider_pref_view

        val a = context.theme.obtainStyledAttributes(
                attrs,
                R.styleable.SliderPreferenceEmbedded,
                0, 0)

        try {
            for (i in 0 until a.indexCount) {
                val attr = a.getIndex(i)

                when (attr) {
                    R.styleable.SliderPreferenceEmbedded_seek_max -> seekBar.max = a.getInteger(attr, 0)
                    R.styleable.SliderPreferenceEmbedded_seek_min -> seekBar.min = a.getInteger(attr, 0)
                    R.styleable.SliderPreferenceEmbedded_scale -> seekBar.scale = a.getFloat(attr, 1f)
                    R.styleable.SliderPreferenceEmbedded_format -> seekBar.format = a.getString(attr)
                    R.styleable.SliderPreferenceEmbedded_show_text -> seekBar.textIndicatorEnabled = a.getBoolean(attr, true)
                    R.styleable.SliderPreferenceEmbedded_discrete_mode -> seekBar.discrete = a.getBoolean(attr, false)

                    R.styleable.SliderPreferenceEmbedded_show_buttons -> {
                        if (!a.getBoolean(attr, true)) {
                            seekBar.up.visibility = View.GONE
                            seekBar.down.visibility = View.GONE
                        }
                    }
                    R.styleable.SliderPreferenceEmbedded_show_reset -> {
                        if (!a.getBoolean(attr, true)) {
                            seekBar.reset.visibility = View.GONE
                        }
                    }
                    R.styleable.SliderPreferenceEmbedded_use_accent_color_for_widget_items -> {
                        if (a.getBoolean(attr, false)) {
                            val colorAttr = context.theme.obtainStyledAttributes(TypedValue().data, intArrayOf(R.attr.colorAccent))
                            val color = colorAttr.getColor(0, 0)
                            colorAttr.recycle()
                            seekBar.up.setColorFilter(color)
                            seekBar.down.setColorFilter(color)
                            seekBar.reset.setColorFilter(color)
                        }
                    }
                }
            }
        } finally {
            a.recycle()
        }
    }

    override fun setDefaultValue(defaultValue: Any?) {
        super.setDefaultValue(defaultValue)
        onSetInitialValue(preferenceManager.sharedPreferences.contains(key), defaultValue)
    }

    override fun onGetDefaultValue(a: TypedArray, index: Int): Int {
        return a.getInt(index, 0)
    }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        seekBar.progress = if (restorePersistedValue) preferenceManager.sharedPreferences.getInt(key, 0) else (defaultValue ?: "0").toString().toInt()
    }

    override fun onCreateView(parent: ViewGroup): View? {
        view = super.onCreateView(parent)
        view.findViewById<LinearLayout>(R.id.seekbar_wrapper).apply {
            seekBar = DiscreteSeekBarText(seekBar)
            addView(seekBar)

            (seekBar.layoutParams as LinearLayout.LayoutParams).apply {
                weight = 1f
                width = LinearLayout.LayoutParams.WRAP_CONTENT
            }
        }

        seekBar.listener = object : OnProgressChangeListener {
            override fun onProgressChanged(seekBar: Slider, fromUser: Boolean, oldPos: Float, newPos: Float, oldValue: Int, newValue: Int) {
                listener?.onPreferenceChange(this@SliderPreferenceEmbedded, newValue)
            }
        }

        viewListener?.viewCreated(this)

        return view
    }

    override fun onBindView(view: View) {
        super.onBindView(view)

        viewListener?.viewBound(this)
    }

    override fun setOnPreferenceChangeListener(onPreferenceChangeListener: Preference.OnPreferenceChangeListener) {
        listener = onPreferenceChangeListener
    }

    interface OnViewCreatedListener {
        fun viewCreated(preferenceEmbeddedNew: SliderPreferenceEmbedded)
        fun viewBound(preferenceEmbeddedNew: SliderPreferenceEmbedded)
    }

    inner class DiscreteSeekBarText constructor(context: Context)
        : LinearLayout(context, null), Slider.OnPositionChangeListener {

        val seekBar: Slider
        val textView: TextView
        val up: ImageView
        val down: ImageView
        val reset: ImageView

        private val prefs = PreferenceManager.getDefaultSharedPreferences(context)

        var scale = 1f //doesn't work for the popup view

        var listener: OnProgressChangeListener? = null

//        var defaultProgress = 0

        var progress: Int
            get() = prefs.getInt(key, -1)
            set(progress) {
                setProgress(progress, true, fromUser)
            }

        var min: Int
            get() = seekBar.minValue
            set(min) {
                seekBar.setValueRange(min, max, false)
                seekBar.setValue(progress.toFloat(), false)
            }

        var max: Int
            get() = seekBar.maxValue
            set(max) {
                seekBar.setValueRange(min, max, false)
                seekBar.setValue(progress.toFloat(), false)
            }

        var textIndicatorEnabled: Boolean
            get() = textView.visibility == View.VISIBLE
            set(enabled) {
                textView.visibility = if (enabled) View.VISIBLE else View.GONE
            }

        var discrete: Boolean
            get() = seekBar.discreteMode
            set(mode) {
                seekBar.discreteMode = mode
            }

        var format: String? = null

        var text: String
            get() = textView.text.toString()
            set(value) {
                val scaled = (value.toDouble() * scale)

                val floatFormat: String

                floatFormat = if (scaled == scaled.toLong().toDouble()) {
                    String.format(Locale.US, "%d", scaled.toLong())
                } else {
                    String.format(Locale.US, "%.2f", scaled)
                }

                if (format == null) {
                    textView.text = if (scale < 1F) floatFormat else scaled.toInt().toString()
                } else {
                    textView.text = String.format(format!!, if (scale < 1F) floatFormat else scaled.toInt().toString())
                }
            }

        private var fromUser = false

        constructor(seekBar: DiscreteSeekBarText) : this(seekBar.context) {
            format = seekBar.format
            scale = seekBar.scale
            listener = seekBar.listener
            progress = seekBar.progress
            min = seekBar.min
            max = seekBar.max
            textIndicatorEnabled = seekBar.textIndicatorEnabled
            discrete = seekBar.discrete
        }

        init {
            View.inflate(context, R.layout.seekbar_with_text, this)

            seekBar = findViewById(R.id.seekbar)
            textView = findViewById(R.id.textview)
            up = findViewById(R.id.up)
            down = findViewById(R.id.down)
            reset = findViewById(R.id.reset)

            seekBar.setOnPositionChangeListener(this)

            val accent = context.theme.obtainStyledAttributes(TypedValue().data, intArrayOf(R.attr.colorAccent))
            val color = accent.getColor(0, 0)
            accent.recycle()

            seekBar.setSecondaryColor(ColorUtils.setAlphaComponent(color, 0x33))

            up.setOnClickListener {
                if (progress < max) setProgress(progress + 1, true, false)
            }

            down.setOnClickListener {
                if (progress > min) setProgress(progress - 1, true, false)
            }

            reset.setOnClickListener {
                resetProgress()
            }
        }

        override fun onPositionChanged(view: Slider, fromUser: Boolean, oldPos: Float, newPos: Float, oldValue: Int, newValue: Int) {
            listener?.onProgressChanged(view, fromUser, oldPos, newPos, oldValue, newValue)

            this.fromUser = fromUser
            if (fromUser) progress = newValue
        }

        fun resetProgress() {
            setProgress(defaultValue.toString().toInt(), true, false)
        }

        fun setProgress(progress: Int, animate: Boolean, fromUser: Boolean) {
            if (!fromUser) seekBar.setValue(progress.toFloat(), animate)
            text = progress.toString()
            prefs.edit().putInt(key, progress).apply()
        }
    }

    interface OnProgressChangeListener {
        fun onProgressChanged(seekBar: Slider, fromUser: Boolean, oldPos: Float, newPos: Float, oldValue: Int, newValue: Int)
    }
}